### PR TITLE
Lesson 3 fix

### PR DESCRIPTION
**Pull request by sergiiromantsov**

### CODE REVIEW

File: lesson-04-memory-management/mm/xxx.c [hunk 1]
```diff
@@ -5,21 +5,142 @@
...
+#define CACHE_SIZE 160
+static struct kmem_cache* g_cache = NULL;
+
+static void cache_constructor( void* p )
+{
+   printk( "%s constructs %p(%u)", THIS_MODULE->name, p, (p ? *(char*)p : 0) );
+}
+
+static int g_mem_config = MEM_CONFIG_KMALLOC;
+module_param_named( mem_config, g_mem_config, int, 0 );
```
Comment (AleksandrBulyshchenko) [inline lesson-04-memory-management/mm/xxx.c:27]: Good idea to use `module_param` for switching implementation, but mixing their code adds a mess.
I'd prefer separate implementations chosen on invocation level or via function table.

File: lesson-04-memory-management/mm/xxx.c [hunk 2]
```diff
@@ -5,21 +5,142 @@
...
+   void* result = NULL;
+   if (!count)
+      return result;
+
+   if (g_mem_config == MEM_CONFIG_KMALLOC)
+   {
+      result = kmalloc( *count, GFP_ATOMIC );
+   }
+   else if (g_mem_config == MEM_CONFIG_KMCACHE)
+   {
```
Comment (AleksandrBulyshchenko) [inline lesson-04-memory-management/mm/xxx.c:61]: Such implementation with `kmem_cache` actually doesn't make much sense - for data which exceed buffer size you just allocate new blocks in a loop and throw away previous one - to implement resizing you'll need to implement some linked list on these blocks.

File: lesson-04-memory-management/mm/xxx.c [hunk 3]
```diff
@@ -5,21 +5,142 @@
...
+      if (g_cache)
+         kmem_cache_free( g_cache, *buffer );
+      *buffer = NULL;
+   }
+}
+
+static size_t construct_buffer( size_t* new_count )
```
Comment (AleksandrBulyshchenko) [inline lesson-04-memory-management/mm/xxx.c]: Why do you pass new_count by pointer if you return modified value?
It just increase code complexity.

File: lesson-04-memory-management/mm/xxx.c [hunk 4]
```diff
@@ -32,11 +153,23 @@ int __init x_init(void) {
...
+      goto error;
+
+   initialize_memory();
+   {
+      char const* const initial_buffer = "Hi!\n";
+      store_to_buffer( initial_buffer, strlen( initial_buffer ) );
+   }
+
+error:
+   printk( "'xxx' module initialized %d\n", res );
```
Comment (AleksandrBulyshchenko) [inline lesson-04-memory-management/mm/xxx.c:171]: There's actually no error handling (like freeing resources)

**Comment by AleksandrBulyshchenko** [comment]
@sergiiromantsov,
Please format the PR according to [wiki](https://github.com/Kernel-GL-HRK/gl-kernel-training-2017/wiki/Pull-Request-Guideline): conform to code standards, add better messages.
The title is not relevant to PR content.

File: lesson-05-time-management/xxxtm.c [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+      seconds = jiffies_to_msecs( diff ) / MSEC_PER_SEC;
+   }
+   sprintf( buf, "Time elapsed since last read: %u seconds\n", seconds );
+   count = strlen( buf );
+   g_last_sys_time = sys_time;
+   printk( "read %ld\n", (long)count );
+   return count;
+}
+
+static ssize_t tm_absk_show( struct class *class, struct class_attribute *attr, char *buf ) {
```
Comment (AleksandrBulyshchenko) [inline lesson-05-time-management/xxxtm.c:30]: This API should return absolute time, not the difference.

File: lesson-05-time-management/xxxtm.c [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+static ssize_t tm_jif_show( struct class *class, struct class_attribute *attr, char *buf ) {
+   size_t count = 0;
+   u64 sys_time = get_jiffies_64();
+   unsigned int seconds = 0;
+   if (g_last_sys_time)
+   {
+      u64 diff = sys_time - g_last_sys_time;
+      seconds = jiffies_to_msecs( diff ) / MSEC_PER_SEC;
+   }
+   sprintf( buf, "Time elapsed since last read: %u seconds\n", seconds );
```
Comment (AleksandrBulyshchenko) [inline lesson-05-time-management/xxxtm.c:23]: Actually, sysfs API not intended for long human-readable strings.

File: lesson-05-time-management/xxxtm.c [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+int __init x_init(void) {
+   int res;
+   tm_jif_class = class_create( THIS_MODULE, "tm_jif-class" );
+   if( IS_ERR( tm_jif_class ) ) printk( "bad class create\n" );
+   res = class_create_file( tm_jif_class, &class_attr_tm_jif );
+
+   tm_absk_class = class_create( THIS_MODULE, "tm_absk-class" );
+   if( IS_ERR( tm_absk_class ) ) printk( "bad class create\n" );
+   res |= class_create_file( tm_absk_class, &class_attr_tm_absk );
+
```
Comment (AleksandrBulyshchenko) [inline lesson-05-time-management/xxxtm.c:66]: What about fibonacci sequence?

File: lesson-04-memory-management/mm/xxx.c (hunk 1 above)
Comment (sergiiromantsov) [reply lesson-04-memory-management/mm/xxx.c:27]: done

File: lesson-04-memory-management/mm/xxx.c (hunk 2 above)
Comment (sergiiromantsov) [reply lesson-04-memory-management/mm/xxx.c:61]: not perfect, but done

File: lesson-04-memory-management/mm/xxx.c (hunk 4 above)
Comment (sergiiromantsov) [reply lesson-04-memory-management/mm/xxx.c:171]: done

**Pull request closed without merging**